src/protocol/libp2p/bitswap/mod.rs

```diff
@@ -147,92 +188,227 @@ impl Bitswap {
-                        multihash_type: cid.hash().code(),
-                        multihash_len: cid.hash().size(),
+        for action in actions {
+            match action {
+                SubstreamAction::SendRequest(cids) => {
+                    if send_request(&mut substream, cids).await.is_err() {
+                        // Drop the substream and all actions in case of sending error.
+                        tracing::debug!(target: LOG_TARGET, ?peer, "bitswap request failed");
+                        return;
                     }
-                    .to_bytes();
+                }
+                SubstreamAction::SendResponse(entries) => {
+                    if send_response(&mut substream, entries).await.is_err() {
+                        // Drop the substream and all actions in case of sending error.
+                        tracing::debug!(target: LOG_TARGET, ?peer, "bitswap response failed");
+                        return;
+                    }
+                }
+            }
+        }
 
-                    response.payload.push(schema::bitswap::Block {
-                        prefix,
-                        data: block,
-                    });
+        self.outbound.insert(peer, substream);
+    }
+
+    /// Handle connection established event.
+    fn on_connection_established(&mut self, peer: PeerId) {
+        // If we have pending actions for this peer, open a substream.
+        if self.pending_dials.remove(&peer) {
+            if let Err(error) = self.service.open_substream(peer) {
+                tracing::debug!(
+                    target: LOG_TARGET,
+                    ?peer,
+                    ?error,
+                    "failed to open substream after connection established",
+                );
+                // Drop all pending actions; they are not going to be handled anyway, and we need
+                // the entry to be empty to properly open subsequent substreams.
+                self.pending_outbound.remove(&peer);
+            }
+        }
+    }
+
+    /// Open substream or dial a peer.
+    fn open_substream_or_dial(&mut self, peer: PeerId) {
+        if let Err(error) = self.service.open_substream(peer) {
+            tracing::trace!(
+                target: LOG_TARGET,
+                ?peer,
+                ?error,
+                "failed to open substream, dialing peer",
+            );
+
+            // Failed to open substream, try to dial the peer.
+            match self.service.dial(&peer) {
+                Ok(()) => {
+                    // Store the peer to open a substream once it is connected.
+                    self.pending_dials.insert(peer);
                 }
-                ResponseType::Presence { cid, presence } => {
-                    response.block_presences.push(schema::bitswap::BlockPresence {
-                        cid: cid.to_bytes(),
-                        r#type: presence as i32,
-                    });
+                Err(error) => {
```

Comment:
nit: We might have an obscure race that have seen in other components/kademlia:
- We are disconnected at T0 on like 351 while `self.service.open_substream(peer)`
- However we had a pending conntions initiated by some other protocol
- By the time we `self.service.dial(&peer)` the peer might have connected
- Maybe its worth for `Err(crate::Error::AlreadyConnected)` to rerequet `self.service.open_substream(peer)`